docs/decisions/0004-error-handling.md

```diff
@@ -0,0 +1,59 @@
+
+- **Exception hierarchy inconsistency**. Half of the custom SK exceptions are derived from SKException, while the other half are directly derived from Exception. This inconsistency in the exception hierarchy does not contribute to a cohesive exception model.
+
+- **Missing original exception details**. Certain SK exceptions do not preserve the original failure or exception details and do not expose them through their properties. This omission prevents SK client code from understanding the problem and handling it properly.
+
+- **No consolidated way for exception handling**. In some cases, SK has an exception type per component implementation, making it impossible for SK client code to handle them in a consolidated manner. Instead of having a single catch block, SK client code needs to include a catch block for each component implementation. Moreover, SK client code needs to be updated every time a new component implementation is added or removed.
+
+## Solution
+
+To address the identified issues, the following steps can be taken:
+
+- **Exception propagation**: To enable SK exceptions to propagate to SK client code, all SK components that currently do not do so should be modified to rethrow the exception instead of suppressing it. This allows SK client code to use try/catch blocks for handling and catching exceptions.
+
+- **Improper exception usage**: Modify SK code to throw .NET standard exceptions, such as ArgumentOutOfRangeException or ArgumentNullException, when class argument values are not provided or are invalid, instead of throwing custom SK exceptions. Analyze SK exception usage to identify other potential areas where standard .NET exceptions can be used instead.
+
+- **Exception hierarchy inconsistency**: Refactor SK custom exceptions that are not derived from SKException to derive from SKException. This ensures consistency in the SK exception hierarchy.
+
+- **Missing original exception details**: Identify all cases where the original exception or failure is not preserved either as an inner exception or as a property/message of the rethrown exception, and address them. For example, AI service connectors that use HttClient should preserve the original status code by either setting it as an inner exception or including it as part of the rethrown exception's property.
+
+- **No consolidated way for exception handling**:
+  - [Memory connectors]: Introduce a new exception called VectorDbStorageException, along with derived exceptions like UpsertRecordException and RemoveRecordException, if additional details need to be captured and propagated to SK client code. Refactor all SK memory component implementations to use these exceptions instead of implementation-specific memory storage exceptions.
```

Comment:
I'm not sure about `VectorDbStorageException` naming. Since we have `IMemoryStore` interface and different implementations (e.g. `ChromaMemoryStore`, `RedisMemoryStore`, I would probably name the exception as `MemoryStoreException`.

Regarding `UpsertRecordException` and `RemoveRecordException`: currently it's possible to use only one memory store per Kernel instance, but in the future, it will be possible to register multiple memory stores. If I register both Chroma and Redis memory stores, when my Kernel instance throws `UpsertRecordException`, how should I know which memory provider throws it? There might be a case, when I want to handle the same operation failure (like upserting record) differently for different memory providers (for example, for MemoryStore1 I want to log error and re-throw, while for MemoryStore2 I want to trace it and continue, since it's expected behavior).

Reply:
MemoryStoreException is definitely a better name for the exception than VectorDbStorageException one. I updated the document to reflect that.

Regarding your second point on handling exceptions from different memory storages differently - Kernel is not the appropriate place for such logic. If Kernel starts distinguishing between memory storages, it will inevitably become coupled to them by referencing the corresponding memory storage specific exception and memory storage projects implicitly.

One possible solution that would enable different handling of upserts elegantly is to use the decorator pattern. The decorator would wrap a specific memory storage and add the required behavior. Since the decorator would implement the same interface as the memory storage, it could seamlessly replace the wrapped memory storage. The Kernel would interact with it without realizing that it is a decorator and not a memory storage.

This approach would prevent the need to modify the memory storage class and avoid cluttering Kernel with memory storage specifics.

Considering that the decorator is designed for a specific memory storage, there is no need to have exceptions that are specific to a particular memory storage implementation. The decorator itself provides enough context for all MemoryStoreException exceptions thrown by the wrapped memory storage.

Reply:
> If Kernel starts distinguishing between memory storages, it will inevitably become coupled to them by referencing the corresponding memory storage specific exception and memory storage projects implicitly.

This is not going to be the case. Kernel only works with `IMemoryStore` interface, Each memory store implementation will be responsible for throwing an exceptions, and Kernel will be responsible to pass it through to the caller. For example, if I have two (potentially in the future) registered memory providers - `ChromaMemoryStore` and `RedisMemoryStore`, then on `kernel.Memory.SaveInformationAsync` in case of exception it will throw `ChromaMemoryStoreException` or `RedisMemoryStoreException` instead of `UpsertRecordException`. I believe there is no need in decorator.

Reply:
Do you see any practical scenarios for the caller of the kernel.Memory.SaveInformationAsync or any other public Kernel method to distinguish between ChromaMemoryStoreException and RedisMemoryStoreException?

Reply:
This can be useful only if we provide a support for multiple memory stores in one Kernel instance (I think there is a plan to add that possibility in the future). In this case I'm not sure how to react to `UpsertRecordException` and what memory provider should be checked.

But even for our current implementation with one memory provider, I'm not sure `UpsertRecordException` will provide some benefits. I think base `MemoryStoreException` with exception message should be enough.